Develop

SabinaBadeeva opened the pull request:
Проект 4го Спринта

SabinaBadeeva committed:
Add gitignore

SabinaBadeeva committed:
Update gitignore

SabinaBadeeva committed:
Добавить тестирование сайта

pomegranate282922 commented on src/main/java/org/example/Main.java:
```diff
@@ -0,0 +1,9 @@
+
+
+
+public class Main {
```
Этот класс лишний, нужно удалить.

pomegranate282922 commented on src/test/java/InputNameSurnameAddressParamTest.java:
```diff
@@ -0,0 +1,78 @@
+        driver.manage().window().maximize();
+        driver.manage().timeouts().pageLoadTimeout(Duration.ofSeconds(10));
+        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
+        driver.get("https://qa-scooter.praktikum-services.ru/order");
```
Урлы лучше выносить как константы, чтобы их можно было переиспользовать

pomegranate282922 commented on src/test/java/InputNameSurnameAddressParamTest.java:
```diff
@@ -0,0 +1,78 @@
+        orderPageTest.setUsername(userName);
+        orderPageTest.setSecondNameField(secondName);
+        orderPageTest.setAddressField(address);
+        Thread.sleep(2000);
```
Такие безусловные ожидания очень тормозят выполнение теста. Нужно заменить на условные ожидания.

pomegranate282922 commented on src/test/java/InputNameSurnameAddressParamTest.java:
```diff
@@ -0,0 +1,78 @@
+        orderPageTest.setAddressField(address);
+        Thread.sleep(2000);
+        orderPageTest.choiceMetroSt();
+        driver.findElement(By.tagName("li")).click();
```
Page object model предполагает, что все локаторы находятся в po.нужно перенести их из теста на уровень po.

pomegranate282922 commented on src/test/java/LogoSamocatTest.java:
```diff
@@ -0,0 +1,48 @@
+            Assert.assertTrue("Element not found", mainUrlPage.contains(logoMain));
+            System.out.println(currentURL + " Мы находимся на главной странице");
+        }
+    }
```
Здесь тоже нужно применить page object model и все локаторы перенести в po.

pomegranate282922 replied:
Не исправлено - локаторов не должно быть в тестах.

pomegranate282922 commented on src/test/java/org/example/QuestionsMainPageTest.java:
```diff
@@ -0,0 +1,85 @@
+            System.out.println(elem.getText());
+        }
+        }
+
```
Нужно реализовать page object model.
нужно проверить все 8 выпадашек.
Эти проверки должны быть полностью независимы друг от друга.

pomegranate282922 replied:
Не исправлено. Оставила комментарий в коде.

pomegranate282922 commented on src/test/java/org/example/OrderPageTest.java:
```diff
@@ -0,0 +1,154 @@
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.concurrent.TimeUnit.valueOf;
+
+public class OrderPageTest {
```
Po не должен содержать слово тест в названии, иначе возникает путаница.
Лучше подобное называть подобным.

pomegranate282922 commented on src/test/java/org/example/LoginTest.java:
```diff
@@ -0,0 +1,86 @@
+        String orderElement = driver.findElement(By.className("Order_ModalHeader__3FDaJ")).getText();
+        String textOrderElement = "Заказ оформлен";
+        Assert.assertTrue("Text not found!", orderElement.contains(textOrderElement));
+
```
Нужно ещё проверить второй флоу, начинающийся с нижней кнопки.

pomegranate282922 reviewed (commented):
Хорошая работа! Реализовала po, тесты написаны и на доп задание тоже.
Есть замечания:
1. Везде нужно убрать локаторы из тестов, перенеся их в po. Там создать методы с действиями и проверками, дав им понятные имена.
2. Доделать тесты на создание заказа ( второй флоу)
3. Сделать все 8 проверок вопросов независимыми друг от друга.
4. Убрать слово тест из имён po.
5. Нужно разложить классы логически по пакетам. Po идут в отдельный пакет в main. Тесты - в test.
6. Остальное - по мелочи.

SabinaBadeeva committed:
Исправить замечания после ревью

pomegranate282922 commented on src/test/java/InputNameSurnameAddressParamTest.java:
```diff
@@ -0,0 +1,80 @@
+            assertEquals( isLoggedPass, ele.getText().contains(metroStation));
+        }
+        if(isLoggedPass){
+            assertEquals(isLoggedPass, driver.findElements(By.xpath(".//*[@id='root']/div/div[2]/div[1]")).size() != 0);
```
Локаторов не должно быть в тестах. Это нарушение page object model.

pomegranate282922 commented on src/test/java/org/example/QuestionsMainPageTest.java:
```diff
@@ -0,0 +1,157 @@
+        //ПРОВЕРЯЕМ, что ЭЛЕМЕНТЫ СОДЕРЖАТ НУЖНЫЙ ТЕКСТ
+        //  1.Элемент СКОЛЬКО ЭТО СТОИТ?
+        mainPage.accordionButtonWhatPriceGetText();
+        String questionWhatPrice = driver.findElement
```
Не должно быть локаторов на уровне тестов. Перенеси все описания локаторов в po.

pomegranate282922 reviewed (commented):
Многое исправила. Супер!
Остались несколько проблемок.
1 есть локаторы на уровне тестов. Это нарушает page object model. Нужно их перенести в po.
2. Остались слипы- их все надо переделать на явные ожидания событий. Иначе они очень замедляют работу тестов.

SabinaBadeeva committed:
Исправить замечания после второго ревью

pomegranate282922 commented on src/main/java/org/example/MainPage.java:
```diff
@@ -0,0 +1,197 @@
+        WebElement element = driver.findElement
+                (orderButton);
+        ((JavascriptExecutor) driver).executeScript("arguments[0].scrollIntoView(true);", element);
+        Thread.sleep(1000);
```
Не убрала слипы. Везде нужно заменить их на условные ожидания соответствующих событий.

pomegranate282922 commented on src/test/java/InputNameSurnameAddressParamTest.java:
```diff
@@ -0,0 +1,74 @@
+        orderPage.setNextButton();
+
+        //Проверяем,что присутствует  элемент метро
+        orderPage.visibleEleMetroStation();
```
Внутри этого метода нет никакой проверки. Просто вывод в консоль. Значит, этот тест ничего не проверяет.

pomegranate282922 commented on src/main/java/org/example/MainPage.java:
```diff
@@ -0,0 +1,197 @@
+        //проверить НАЛИЧИЕ ВСЕХ ВОПРОСОВ
+        List<WebElement> allElements = driver.findElements(listAllQuestions);
+        for (WebElement elem : allElements) {
+            System.out.println(elem.getText());
```
Здесь нет проверки- просто вывод в консоль. Нужна проверка ( то есть тест должен падать, если условие не выполняется).

pomegranate282922 commented on src/main/java/LogoSamocat.java:
```diff
@@ -0,0 +1,36 @@
+    public void isDisplayedClickLogo() {
+        WebElement logoPresent = driver.findElement(headerLogoScooter);
+        if (driver.findElement(By.cssSelector("a.Header_LogoScooter__3lsAR > img")).isDisplayed()) {
+            driver.findElement(By.cssSelector("a.Header_LogoScooter__3lsAR > img")).click();}
```
Здесь нет проверки. То есть если Элемента нет, он не нажмется. Но тест при этом останется зелёным.

pomegranate282922 commented on src/main/java/org/example/MainPage.java:
```diff
@@ -0,0 +1,197 @@
+    // 1.метод для элемента "СКОЛЬКО ЭТО СТОИТ?"
+    public void clickAccordionButtonWhatPrice(String dataUpperText){
+         driver.findElement(accordionButtonWhatPrice).click();
+        driver.findElement(accordionButtonWhatPrice).getText();
```
Везде берётся текст у элемента и нигде не используется. Можно убрать ненужный код.

pomegranate282922 commented on src/main/java/org/example/OrderPage.java:
```diff
@@ -0,0 +1,210 @@
+        if(i != 0)
+            System.out.println(i);
+        else
+            System.out.println("Element is not present");
```
Здесь тоже нет проверки.

pomegranate282922 reviewed (commented):
Все же требуется ещё доработка.
1. Нужно убрать все слипы из кода 
2. Каждый тест должен содержать проверку. Не просто вывод чего-то в консоль, а проверку, то есть падать, если условие не выполняется.

SabinaBadeeva committed:
Убрать слипы, исправить проверки

pomegranate282922 commented on src/main/java/org/example/MainPage.java:
```diff
@@ -0,0 +1,175 @@
+        new WebDriverWait(driver, Duration.ofSeconds(5))
+                .until(ExpectedConditions.textToBePresentInElementLocated(accordionButtonWantSome, dataUpperTextTwo));}
+    public void getAnswerButtonWantSome(String dataTextTwo) {
+        driver.findElement(answerButtonWantSome).getText();
```
Ненужная строка

pomegranate282922 commented on src/main/java/org/example/MainPage.java:
```diff
@@ -0,0 +1,175 @@
+        new WebDriverWait(driver, Duration.ofSeconds(5))
+                .until(ExpectedConditions.textToBePresentInElementLocated(accordionButtonTimeRental, dataUpperTextThree));}
+    public void getAnswerButtonTimeRental(String dataTextThree) {
+        driver.findElement(answerAccordionButtonTimeRental).getText();
```
Ненужная строка.
Везде аналогично.

pomegranate282922 commented on src/main/java/org/example/MainPage.java:
```diff
@@ -0,0 +1,175 @@
+                .until(ExpectedConditions.textToBePresentInElementLocated(answerButtonLiveFar, dataTextEight));}
+
+    //метод вывода наличия вопросов
+    public String allElemQuestion() {
```
Логика этого метода непонятна. 
ты выводишь первый из элементов списка.
Если вдруг список окажется пустым, метод будет вызывать самого себя и зациклится навечно.

pomegranate282922 commented on src/test/java/InputNameSurnameAddressParamTest.java:
```diff
@@ -0,0 +1,91 @@
+
+        // Проверяем, что виден элемент с заголовком "ПРО АРЕНДУ" на следующей странице, если регистрация прошла
+        // если данные введены неверно (регистрация не прошла), тест падает
+        String header = orderPage.isDisplayedHeader();
```
Нет, логика теста должна быть другой. 
Если поведение ожидаемое, то тест не падает.
Если поведение не ожидаемое, то тест падает.

pomegranate282922 commented on src/test/java/org/example/OrderParamTest.java:
```diff
@@ -0,0 +1,100 @@
+
+        // проверка, что видна надпись в Поп-Ап,если заказ оформлен
+        String order = orderPage.isDisplayedHeader();
+        assertTrue(order, isOrderPass);
```
Здесь тоже нужно переделать логику. Зачем нужен параметр isOrderPass , если он всегда true. Данный ассерт всегда проходит, значит, он не нужен.

pomegranate282922 commented on src/test/java/org/example/QuestionsMainPageTest.java:
```diff
@@ -0,0 +1,136 @@
+        //проверить, что в МАССИВЕ с ВОПРОСАМИ действительно содержится один из вопросов
+        String allQuestions =  mainPage.allElemQuestion();
+        String someQuestion =  "Можно ли отменить заказ?";
+       assertTrue(allQuestions.contains(someQuestion));
```
Здесь нужно поменять логику метода ( написала в po). Но в принципе, сам тест кажется избыточен.

pomegranate282922 reviewed (commented):
Многое поправила, но есть ещё необходимость доработок.

SabinaBadeeva committed:
Сделала, как просят по тестовому сценарию в задании:
 - два позитивных теста с разными наборами данных, с двумя точками входа.
 Проверила,что появилось всплывающее окно с сообщением об успешном создании заказа.
 В разделе «Вопросы о важном» проверила -  когда нажимаешь на стрелочку, открывается соответствующий текст.
  Убрала негативный тест и дополнительную проверку,которая не требуется в задании.

pomegranate282922 reviewed (approved):
Зачет!

SabinaBadeeva committed:
Update README.md

SabinaBadeeva committed:
Update QuestionsMainPageTest.java
